extensions/replication/tasks/CopyLocationTask.js

```diff
@@ -0,0 +1,794 @@
+            if (res.error) {
+                log.error('error parsing metadata blob', Object.assign({
+                    error: res.error,
+                    method: 'CopyLocationTask._getSourceMD',
+                }, actionEntry.getLogInfo()));
+                return cb(errors.InternalError.
+                    customizeDescription('error parsing metadata blob'));
+            }
+            return cb(null, res.result);
+        });
+    }
+
+    _getAndPutObject(actionEntry, objMD, log, cb) {
+        const partLogger = this.logger.newRequestLogger(log.getUids());
+        const extMetrics = _getExtMetrics(this.site,
+            objMD.getContentLength(), actionEntry);
+        this.mProducer.publishMetrics(extMetrics,
+            metricsTypeQueued, metricsExtension, () => {});
+        this.retry({
+            actionDesc: 'stream part data',
+            logFields: { entry: actionEntry.getLogInfo() },
+            actionFunc: done => this._getAndPutObjectOnce(
+                actionEntry, objMD, partLogger, done),
+            shouldRetryFunc: err => err.retryable,
+            log: partLogger,
+        }, cb);
+    }
+
+    _getAndPutObjectOnce(actionEntry, objMD, log, done) {
+        log.debug('getting object data', actionEntry.getLogInfo());
+        const doneOnce = jsutil.once(done);
+        const size = objMD.getContentLength();
+        let incomingMsg = null;
+        if (size !== 0) {
+            const { bucket, key, version } = actionEntry.getAttribute('target');
+            const sourceReq = this.backbeatClient.getObject({
+                Bucket: bucket,
+                Key: key,
+                VersionId: version,
+                LocationConstraint: objMD.getDataStoreName(),
+            });
+            attachReqUids(sourceReq, log);
+            sourceReq.on('error', err => {
+                // eslint-disable-next-line no-param-reassign
+                if (err.statusCode === 404) {
+                    log.error('the source object was not found', Object.assign({
+                        method: 'CopyLocationTask._getAndPutObjectOnce',
+                        peer: this.sourceConfig.s3,
+                        error: err.message,
+                        httpStatus: err.statusCode,
+                    }, actionEntry.getLogInfo()));
+                    return doneOnce(err);
+                }
+                log.error('an error occurred on getObject from S3',
+                    Object.assign({
+                        method: 'CopyLocationTask._getAndPutObjectOnce',
+                        peer: this.sourceConfig.s3,
+                        error: err.message,
+                        httpStatus: err.statusCode,
+                    }, actionEntry.getLogInfo()));
+                return doneOnce(err);
+            });
+            incomingMsg = sourceReq.createReadStream();
+            incomingMsg.on('error', err => {
+                if (err.statusCode === 404) {
```

Comment:
Just curious here: does this error occur if the object is deleted mid-stream?

Reply:
I don't think so, I would expect some 50x, it likely depends on the backend type too, but it would be a good test to do at least once to know the behavior in such case (although hard to produce and even harder to automate...). But in any case it should lead to a failure that does not trigger a change on the source metadata.